lib/galaxy/jobs/runners/drmaauniva.py

```diff
@@ -0,0 +1,640 @@
+"""
+Kind of fail save job control via the DRMAA API.
+Tested for UNIVA grid engine
+"""
+import logging
+import os
+import re
+import signal
+import subprocess
+import time
+
+# from galaxy import model
+from galaxy.jobs.runners.drmaa import DRMAAJobRunner
+
+
+log = logging.getLogger( __name__ )
+
+__all__ = ( 'DRMAAUnivaJobRunner', )
+
+MEMORY_LIMIT_SCAN_SIZE = 1024 * 1024  # 1MB
+
+"""
+TODO
+find out what it means if there is no error code file (.ec)
+can this also be used
+
+tool wrapper:
+- A tool that is the bash script itself.
+- bash script (create long array, dd into variable)
+- C
+- C++
+- python
+- java
+- perl
+"""
+
+
+class DRMAAWaitUnusable( Exception ):
+    pass
+
+
+class DRMAAUnivaJobRunner( DRMAAJobRunner ):
+    runner_name = "DRMAAUnivaJobRunner"
+    # restrict job name length as in the DRMAAJobRunner
+    # restrict_job_name_length = 15
+
+    def check_watched_item( self, ajs, new_watched ):
+        # get state with job_info/qstat
+        state = self._get_drmaa_state(ajs.job_id, self.ds, False)
+        logging.debug("DRMAAUnivaJobRunner:check_watched_item ({jobid}) -> state {state}".format(jobid=ajs.job_id, state=self.drmaa_job_state_strings[state]))
+        return state
+
+    def _complete_terminal_job( self, ajs, drmaa_state, **kwargs ):
+        extinfo = dict()
+        # get state with job_info/qstat + wait/qacct
+        state = self._get_drmaa_state(ajs.job_id, self.ds, True, extinfo)
+        logging.debug("DRMAAUnivaJobRunner:_complete_terminal_job ({jobid}) -> state {state} info {info}".format(jobid=ajs.job_id, state=self.drmaa_job_state_strings[state], info=extinfo))
+
+        # for any job that is finished (regardless of successful or failed) check if there
+        # were memory or time deletions or if the job was deleted
+        if state in [ self.drmaa.JobState.DONE, self.drmaa.JobState.FAILED ]:
+            # get configured job destination
+            job_destination = ajs.job_wrapper.job_destination
+            native_spec = job_destination.params.get('nativeSpecification', None)
+            # determine time and memory that was granted for the job
+            time_granted, mem_granted = _parse_native_specs( ajs.job_id, native_spec )
+            time_wasted = extinfo["time_wasted"]
+            mem_wasted = extinfo["memory_wasted"]
+
+            # check if the output contains indicators for a memory violation
+            memviolation = _check_memory_limit( ajs.error_file )
+            logging.debug("DRMAAUnivaJobRunner:_complete_terminal_job ({jobid}) memviolation {mv}".format(jobid=ajs.job_id, mv=memviolation))
+
+            # check job for run time or memory violation
+            if "deleted" in extinfo and extinfo[ "deleted" ]:
```

Comment:
Whitespace, should be `if "deleted" in extinfo and extinfo["deleted"]:`

Reply:
done